### PR TITLE
[feature] add support for StreamNative OAuth2 authentication for pulsar.

### LINKED ISSUE #12067: [feature] Add support for oauth2 to Pulsar ingest plugin
StreamNative (the cloud SAAS offering of Pulsar) uses OAuth2 to authenticate clients to their Pulsar clusters. The current Pulsar ingestion accepts a JWT as authentication, but cannot be setup to use OAuth2.  I also have a PR to add this support.

### PR DESCRIPTION
**Pull request by JeffBolle**
Adds support for StreamNative OAuth2 authentication, allowing Pinot to connect to StreamNative hosted Pulsar cluster.

fixes #12067

### CODE REVIEW

**Comment by Jackie-Jiang** [comment]
Thanks for the contribution! @KKcorps Can you help take a look?

File: pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/main/java/org/apache/pinot/plugin/stream/pulsar/PulsarPartitionLevelConnectionHandler.java [hunk 1]
```diff
@@ -60,13 +65,34 @@ public PulsarPartitionLevelConnectionHandler(String clientId, StreamConfig strea
...
+  private Optional<Authentication> getAuthenticationFactory(PulsarConfig pulsarConfig) {
+    if (StringUtils.isNotBlank(pulsarConfig.getIssuerUrl())
+        && StringUtils.isBlank(pulsarConfig.getAudience())
+        && StringUtils.isBlank(pulsarConfig.getCredentialsFilePath())) {
+      try {
+        return Optional.of(AuthenticationFactoryOAuth2.clientCredentials(
+            new URL(pulsarConfig.getIssuerUrl()),
+            new URL(pulsarConfig.getCredentialsFilePath()),
+            pulsarConfig.getAudience()));
+      } catch (MalformedURLException mue) {
```
Comment (KKcorps) [inline pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/main/java/org/apache/pinot/plugin/stream/pulsar/PulsarPartitionLevelConnectionHandler.java:85]: Do we need to handle any other exceptions in this block?

File: pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/main/java/org/apache/pinot/plugin/stream/pulsar/PulsarConfig.java [hunk 2]
```diff
@@ -52,45 +56,50 @@ public class PulsarConfig {
   private final SubscriptionInitialPosition _subscriptionInitialPosition;
   private final String _authenticationToken;
   private final String _tlsTrustCertsFilePath;
+
+  private final String _issuerUrl;
+  private final String _credentialsFilePath; // Absolute path of your downloaded key file
```
Comment (KKcorps) [inline pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/main/java/org/apache/pinot/plugin/stream/pulsar/PulsarConfig.java:61]: How will someone provide this file in pinot deployments? I think if it's running in docker, expecting folks to copy the credentials file to docker volumn is wrong choice.

Comment (KKcorps) [reply pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/main/java/org/apache/pinot/plugin/stream/pulsar/PulsarConfig.java:61]: Is there a way we can mention encrypted version of this in table config or mention something like S3 path?

File: pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/main/java/org/apache/pinot/plugin/stream/pulsar/PulsarConfig.java [hunk 3]
```diff
@@ -52,45 +56,50 @@ public class PulsarConfig {
   private final SubscriptionInitialPosition _subscriptionInitialPosition;
   private final String _authenticationToken;
   private final String _tlsTrustCertsFilePath;
+
+  private final String _issuerUrl;
```
Comment (KKcorps) [inline pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/main/java/org/apache/pinot/plugin/stream/pulsar/PulsarConfig.java:60]: Add comment on what ech of these new configs mean along with sample values.

File: pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/main/java/org/apache/pinot/plugin/stream/pulsar/PulsarConfig.java (hunk 2 above)
Comment (KKcorps) [reply pinot-plugins/pinot-stream-ingestion/pinot-pulsar/src/main/java/org/apache/pinot/plugin/stream/pulsar/PulsarConfig.java]: Ok,in that case mention a small TODO on top of it that mentions to find a good way to push this file to all servers

**Review by KKcorps — APPROVED** [review]
LGTM! just one minor comment

**Pull request merged**